File: scripts/clean-quran-sql.js
#!/usr/bin/env node
/**
 * ⚠️ OBSOLETE — kept only as a record of how the clean file was produced.
 *
 * This generated hafsData_v2-0-clean.sql from the raw hafsData_v2-0.sql by
 * stripping end-of-ayah markers (ﰀ ﰁ ﰂ — U+FC00..U+FDFF), which are typography
 * rather than Quran text.
 *
 * The raw file is no longer kept in the repository: the clean file is now the
 * canonical source and is committed directly. Running this script will fail on
 * the missing input, which is expected. Do not "fix" it by pointing it at the
 * clean file — that would just rewrite the source onto itself.
 */

const fs = require('fs');
const path = require('path');

const inputFile = path.join(__dirname, '..', 'hafsData_v2-0.sql');
const outputFile = path.join(__dirname, '..', 'hafsData_v2-0-clean.sql');

console.log('🔄 Reading SQL file...');

try {
  // Read the file
  const content = fs.readFileSync(inputFile, 'utf8');
  
  console.log(`📄 Original file: ${content.split('\n').length} lines`);
  
  // Define regex to match end-of-ayah markers
  // These are in Unicode range U+FC00 to U+FDFF (Arabic Presentation Forms-A)
  const ayahMarkersRegex = /[\uFC00-\uFDFF]/g;
  
  // Count markers before removal
  const markers = content.match(ayahMarkersRegex);
  const markersCount = markers ? markers.length : 0;
  
  console.log(`🔍 Found ${markersCount} end-of-ayah markers to remove`);
  
  // Remove the markers
  let cleanedContent = content.replace(ayahMarkersRegex, '');
  
  // Remove trailing spaces in Arabic text fields (before ','<text>')
  // This removes spaces at the end of aya_text field
  cleanedContent = cleanedContent.replace(/\s+','([^']+)'\)/g, "','$1')");
  
  console.log(`🧹 Cleaned trailing spaces from ayah text`);
  
  // Write cleaned content
  console.log(`💾 Writing cleaned file to: ${outputFile}`);
  fs.writeFileSync(outputFile, cleanedContent, 'utf8');
  
  console.log(`✅ Success! Cleaned ${markersCount} ayah markers`);
  console.log(`📝 Clean file saved: hafsData_v2-0-clean.sql`);
  
} catch (error) {
  console.error('❌ Error:', error.message);
  process.exit(1);
}

File: scripts/clean-quran-sql.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Script to clean Quran SQL file by removing end-of-ayah markers
These Unicode symbols (ﰀ ﰁ ﰂ etc.) are not part of the Quran text
"""

import re
import sys

def clean_sql_file(input_file, output_file):
    """Remove end-of-ayah markers from SQL file"""
    
    # Define the Unicode range for end-of-ayah markers
    # These are in the Arabic Presentation Forms-A block (U+FB50 to U+FDFF)
    # Specifically U+FC00 onwards
    ayah_markers_pattern = r'[\uFC00-\uFDFF]'
    
    print(f"Reading {input_file}...")
    
    try:
        with open(input_file, 'r', encoding='utf-8') as f:
            content = f.read()
        
        original_lines = content.count('\n')
        print(f"Original file: {original_lines} lines")
        
        # Remove the ayah markers
        cleaned_content = re.sub(ayah_markers_pattern, '', content)
        
        # Count how many markers were removed
        markers_removed = len(re.findall(ayah_markers_pattern, content))
        print(f"Removed {markers_removed} end-of-ayah markers")
        
        # Write cleaned content
        print(f"Writing to {output_file}...")
        with open(output_file, 'w', encoding='utf-8') as f:
            f.write(cleaned_content)
        
        print(f"✅ Success! Cleaned file saved to: {output_file}")
        print(f"Total ayahs cleaned: {markers_removed}")
        
    except Exception as e:
        print(f"❌ Error: {e}")
        sys.exit(1)

if __name__ == "__main__":
    input_file = r"e:\Halim\freelancer\Quran-Yutla\hafsData_v2-0.sql"
    output_file = r"e:\Halim\freelancer\Quran-Yutla\hafsData_v2-0-clean.sql"
    
    clean_sql_file(input_file, output_file)
